topydo/lib/EditCommand.py

```diff
@@ -15,27 +15,121 @@
     def __init__(self, p_args, p_todolist, p_output, p_error, p_input):
         super(EditCommand, self).__init__(p_args, p_todolist, p_output,
             p_error, p_input)
 
+        self.is_expression = False
+
+    def _process_flags(self):
+        opts, args = self.getopt('xe')
+
+        for opt, value in opts:
+            if opt == '-x':
+                self.show_all = True
+            elif opt == '-e':
+                self.is_expression = True
+
+        self.args = args
+
+    def _todos_to_temp(self):
+        f = tempfile.NamedTemporaryFile()
+        for todo in self.todos:
+            f.write("%s\n" % todo.__str__())
+        f.seek(0)
+
+        return f
+
+    def _todos_from_temp(self, temp_file):
+        temp_file.seek(0)
+        todos = temp_file.read().splitlines()
+
+        todo_objs = []
+        for todo in todos:
+            todo_objs.append(Todo(todo))
+
+        return todo_objs
+
+    def _open_in_editor(self, temp_file, editor):
+        try:
+            return check_call([editor, temp_file.name])
+        except(CalledProcessError):
+            self.error('Something went wrong in the editor...')
+            return 1
+
+    def _catch_todo_errors(self):
+        errors = []
+
+        if len(self.invalid_numbers) > 1 or len(self.invalid_numbers) > 0 and len(self.todos) > 0:
+            for number in self.invalid_numbers:
+                errors.append("Invalid todo number given: {}.".format(number))
+        elif len(self.invalid_numbers) == 1 and len(self.todos) == 0:
+            errors.append("Invalid todo number given.")
+
+        if len(errors) > 0:
+            return errors
+        else:
+            return None
+
     def execute(self):
-        if not super(EditCommand, self).execute():
+        if not super(ListCommand, self).execute():
             return False
 
+        self.printer.add_filter(PrettyPrinterNumbers(self.todolist))
         try:
             editor = os.environ['EDITOR'] or 'vi'
         except(KeyError):
             editor =  'vi'
-        todo = config().todotxt()
 
-        return call([editor, todo]) == 0
+        try:
+            if len(self.args) < 1:
+                todo = config().todotxt()
+
+                return call([editor, todo]) == 0
+            else:
+                self._process_flags()
+
+                if self.is_expression:
+                    self.todos = self._view()._viewdata
+                else:
+                    self.get_todos(self.args)
+
+                todo_errors = self._catch_todo_errors()
+
+                if not todo_errors:
+                    temp_todos = self._todos_to_temp()
+
+                    if not self._open_in_editor(temp_todos, editor):
+                        new_todos = self._todos_from_temp(temp_todos)
+                        if len(new_todos) == len(self.todos):
+                            for todo in self.todos:
+                                self.todolist.delete(todo)
```

Comment:
Are dependencies between todo items properly preserved? Might need a test case to check this.